modules/pymol/shortcut.py

```diff
@@ -1,184 +1,203 @@
-                            msg += " Choices:\n"
-                            msg += "\n".join(lst)
-                    raise parsing.QuietException(msg)
-
-            else:
-                result = self.interpret(kee)
-                if not is_string(result):
-                    if descrip is not None:
-                        lst = parsing.list_to_str_list(result)
-                        msg = "Error: ambiguous %s:\n%s" % (descrip, '\n'.join(lst))
-                        raise parsing.QuietException(msg)
+# A* -------------------------------------------------------------------
+# B* This file contains source code for the PyMOL computer program
+# C* Copyright (c) Schrodinger, LLC.
+# D* -------------------------------------------------------------------
+# E* It is unlawful to modify or remove this copyright notice.
+# F* -------------------------------------------------------------------
+# G* Please see the accompanying LICENSE file for further information.
+# H* -------------------------------------------------------------------
+# I* Additional authors of this source file include:
+# -*
+# -*
+# -*
+# Z* -------------------------------------------------------------------
+
+from typing import Iterable, Optional
+from collections import defaultdict
+from pymol import parsing
+
+
+class Shortcut:
+    def __init__(
+        self,
+        keywords: Optional[Iterable] = None,
+        filter_leading_underscore: bool = True,
+    ):
+        keywords = list(keywords) if keywords is not None else []
+        self.filter_leading_underscore = filter_leading_underscore
+        self.keywords = (
+            [keyword for keyword in keywords if keyword[:1] != "_"]
+            if filter_leading_underscore
+            else keywords
+        )
+        self.shortcut: dict[str, str | int] = {}
+        self.abbreviation_dict = defaultdict(list)
+
+        for keyword in self.keywords:
+            self.optimize_symbols(keyword)
+
+        self._rebuild_finalize()
+
+    def __contains__(self, keyword: str) -> bool:
+        return keyword in self.shortcut
+
+    def __getitem__(self, keyword: str) -> Optional[int | str]:
+        return self.shortcut.get(keyword)
+
+    def __delitem__(self, keyword: str) -> None:
+        self.keywords.remove(keyword)
+        self.rebuild()
+
+    def _make_abbreviation(self, s: str, groups_length: int) -> str:
+        """
+        Creates an abbreviation for a string by shortening its components.
+        The abbreviation takes the first `groups_length`
+        characters of each part before the last component.
+
+        Example 1:
+        Input: s:'abc_def_ghig', groups_length: 1
+        Output: 'a_d_ghig'
+
+        Example 2:
+        Input: s:'abc_def', groups_length: 2
+        Output: 'ab_def'
+        """
+        groups = s.split("_")
+        groups[:-1] = [c[0:groups_length] for c in groups[:-1]]
+        return "_".join(groups)
+
+    def optimize_symbols(self, keyword: str) -> None:
```

Comment:
Same for this one; probably just for internal usage.

Reply:
done